File: lib/services/csv_service.dart
import 'package:flutter/services.dart';
import 'package:csv/csv.dart';
import '../models/tufe_data.dart';
import 'github_csv_service.dart';

class CSVService {
  static Future<List<TufeData>> loadTufeData() async {
    try {
      final String csvData = await GitHubCSVService.loadCSVFromGitHub(
          'gruplaraylik.csv',
          useCache: false);
      print('CSV parse ediliyor... Uzunluk: ${csvData.length}');

      // Manuel olarak satırlara böl
      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      print('Manuel satır sayısı: ${lines.length}');

      // Her satırı ayrı ayrı CSV olarak parse et
      List<List<dynamic>> rows = [];
      for (int i = 0; i < lines.length; i++) {
        String line = lines[i].trim();
        if (line.isNotEmpty) {
          try {
            List<List<dynamic>> parsedLine =
                const CsvToListConverter().convert(line);
            if (parsedLine.isNotEmpty) {
              rows.add(parsedLine[0]);
              print('Satır $i parse edildi: ${parsedLine[0].length} sütun');
            }
          } catch (e) {
            print('Satır $i parse hatası: $e');
          }
        }
      }

      print('Başarıyla parse edilen satır sayısı: ${rows.length}');

      if (rows.isNotEmpty) {
        print('Header satırı: ${rows[0]}');
        if (rows.length > 1) {
          print('İlk veri satırı: ${rows[1]}');
        }
      }

      if (rows.isEmpty) {
        throw Exception('CSV dosyası boş');
      }

      List<TufeData> tufeDataList = [];

      // İlk satır header, ondan sonraki satırları işle
      for (int i = 1; i < rows.length; i++) {
        if (rows[i].length < 2) {
          print('Geçersiz satır $i: ${rows[i]}');
          continue; // Geçersiz satırları atla
        }

        String groupName = rows[i][1].toString().trim();
        print('Satır $i - Grup: "$groupName", Uzunluk: ${rows[i].length}');

        if (groupName.isEmpty) {
          print('Boş grup adı, atlaniyor');
          continue;
        }

        // Son sütundaki veriyi al (en güncel ay)
        dynamic lastValue = rows[i][rows[i].length - 1];
        double changeRate = 0.0;

        if (lastValue is num) {
          changeRate = lastValue.toDouble();
        } else if (lastValue is String) {
          try {
            changeRate = double.parse(lastValue);
          } catch (e) {
            // Parse edilemeyen değerleri 0 olarak kabul et
            changeRate = 0.0;
          }
        }

        tufeDataList.add(TufeData(
          groupName: groupName,
          changeRate: changeRate,
        ));
        print('Eklendi: $groupName = $changeRate');
      }

      print('Toplam ${tufeDataList.length} veri eklendi');

      // Verileri değişim oranına göre büyükten küçüğe sırala
      tufeDataList.sort((a, b) => b.changeRate.compareTo(a.changeRate));

      print('Sıralama sonrası ilk 3 item:');
      for (int i = 0; i < tufeDataList.length && i < 3; i++) {
        print(
            '${i + 1}. ${tufeDataList[i].groupName}: ${tufeDataList[i].changeRate}');
      }

      return tufeDataList;
    } catch (e) {
      throw Exception('CSV veri yükleme hatası: $e');
    }
  }

  static String getCurrentMonth() {
    final now = DateTime.now();
    const months = [
      'Ocak',
      'Şubat',
      'Mart',
      'Nisan',
      'Mayıs',
      'Haziran',
      'Temmuz',
      'Ağustos',
      'Eylül',
      'Ekim',
      'Kasım',
      'Aralık'
    ];
    return months[now.month - 1];
  }

  static Future<String> getMonthFromCSV() async {
    try {
      final String csvData = await GitHubCSVService.loadCSVFromGitHub(
          'gruplaraylik.csv',
          useCache: false);

      // Manuel olarak satırlara böl
      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      List<List<dynamic>> rows = [];

      for (String line in lines) {
        String trimmedLine = line.trim();
        if (trimmedLine.isNotEmpty) {
          try {
            List<List<dynamic>> parsedLine =
                const CsvToListConverter().convert(trimmedLine);
            if (parsedLine.isNotEmpty) {
              rows.add(parsedLine[0]);
            }
          } catch (e) {
            // Parse hatalarını sessizce atla
          }
        }
      }

      if (rows.isEmpty || rows[0].length < 3) {
        // Fallback olarak sistem tarihini kullan
        return getCurrentMonth();
      }

      // Header satırından son sütundaki tarihi al
      String lastDateString = rows[0].last.toString().trim();
      
      // Tarih formatını parse et (2025-08-31 formatında)
      try {
        DateTime lastDate = DateTime.parse(lastDateString);
        const months = [
          'Ocak',
          'Şubat',
          'Mart',
          'Nisan',
          'Mayıs',
          'Haziran',
          'Temmuz',
          'Ağustos',
          'Eylül',
          'Ekim',
          'Kasım',
          'Aralık'
        ];
        return months[lastDate.month - 1];
      } catch (e) {
        // Tarih parse edilemezse sistem tarihini kullan
        return getCurrentMonth();
      }
    } catch (e) {
      // Hata durumunda sistem tarihini kullan
      return getCurrentMonth();
    }
  }

  static Future<double> getTufeMonthlyChange() async {
    try {
      final String csvData = await GitHubCSVService.loadCSVFromGitHub(
          'gruplaraylik.csv',
          useCache: false);

      // Manuel olarak satırlara böl
      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      List<List<dynamic>> rows = [];

      for (String line in lines) {
        String trimmedLine = line.trim();
        if (trimmedLine.isNotEmpty) {
          try {
            List<List<dynamic>> parsedLine =
                const CsvToListConverter().convert(trimmedLine);
            if (parsedLine.isNotEmpty) {
              rows.add(parsedLine[0]);
            }
          } catch (e) {
            // Parse hatalarını sessizce atla
          }
        }
      }

      if (rows.isEmpty) {
        throw Exception('CSV dosyası boş');
      }

      // Find the "Web TÜFE" row
      for (int i = 1; i < rows.length; i++) {
        if (rows[i].length >= 2) {
          final String groupName = rows[i][1].toString().trim();
          if (groupName == 'Web TÜFE') {
            // Get the last column (latest monthly change)
            final String lastValue = rows[i].last.toString().trim();
            try {
              return double.parse(lastValue);
            } catch (e) {
              return 0.0;
            }
          }
        }
      }

      return 0.0;
    } catch (e) {
      return 0.0;
    }
  }

  /// CSV dosyasındaki tüm tarihleri döndürür
  static Future<List<String>> getAvailableDates() async {
    try {
      final String csvData = await GitHubCSVService.loadCSVFromGitHub(
          'gruplaraylik.csv',
          useCache: false);

      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      List<List<dynamic>> rows = [];

      for (String line in lines) {
        String trimmedLine = line.trim();
        if (trimmedLine.isNotEmpty) {
          try {
            List<List<dynamic>> parsedLine =
                const CsvToListConverter().convert(trimmedLine);
            if (parsedLine.isNotEmpty) {
              rows.add(parsedLine[0]);
            }
          } catch (e) {
            // Parse hatalarını sessizce atla
          }
        }
      }

      if (rows.isEmpty || rows[0].length < 3) {
        return [];
      }

      // Header satırından tarihleri al (2. sütundan itibaren)
      List<String> dates = [];
      for (int i = 2; i < rows[0].length; i++) {
        String dateStr = rows[0][i].toString().trim();
        if (dateStr.isNotEmpty) {
          dates.add(dateStr);
        }
      }

      return dates;
    } catch (e) {
      print('Tarih listesi yükleme hatası: $e');
      return [];
    }
  }

  /// Belirli bir tarih için TÜFE verilerini yükler
  static Future<List<TufeData>> loadTufeDataForDate(String selectedDate) async {
    try {
      final String csvData = await GitHubCSVService.loadCSVFromGitHub(
          'gruplaraylik.csv',
          useCache: false);

      List<String> lines = csvData.split(RegExp(r'\r?\n'));
      List<List<dynamic>> rows = [];

      for (int i = 0; i < lines.length; i++) {
        String line = lines[i].trim();
        if (line.isNotEmpty) {
          try {
            List<List<dynamic>> parsedLine =
                const CsvToListConverter().convert(line);
            if (parsedLine.isNotEmpty) {
              rows.add(parsedLine[0]);
            }
          } catch (e) {
            print('Satır $i parse hatası: $e');
          }
        }
      }

      if (rows.isEmpty) {
        throw Exception('CSV dosyası boş');
      }

      // Header satırından seçili tarihin sütun indeksini bul
      int dateColumnIndex = -1;
      for (int i = 2; i < rows[0].length; i++) {
        String dateStr = rows[0][i].toString().trim();
        if (dateStr == selectedDate) {
          dateColumnIndex = i;
          break;
        }
      }

      if (dateColumnIndex == -1) {
        throw Exception('Seçili tarih bulunamadı: $selectedDate');
      }

      List<TufeData> tufeDataList = [];

      // İlk satır header, ondan sonraki satırları işle
      for (int i = 1; i < rows.length; i++) {
        if (rows[i].length < dateColumnIndex + 1) {
          continue;
        }

        String groupName = rows[i][1].toString().trim();
        if (groupName.isEmpty) {
          continue;
        }

        // Seçili tarih sütunundaki veriyi al
        dynamic value = rows[i][dateColumnIndex];
        double changeRate = 0.0;

        if (value is num) {
          changeRate = value.toDouble();
        } else if (value is String) {
          try {
            changeRate = double.parse(value);
          } catch (e) {
            changeRate = 0.0;
          }
        }

        tufeDataList.add(TufeData(
          groupName: groupName,
          changeRate: changeRate,
        ));
      }

      // Verileri değişim oranına göre büyükten küçüğe sırala
      tufeDataList.sort((a, b) => b.changeRate.compareTo(a.changeRate));

      return tufeDataList;
    } catch (e) {
      throw Exception('CSV veri yükleme hatası: $e');
    }
  }

  /// Belirli bir tarih için ay adını döndürür
  static String getMonthFromDate(String dateString) {
    try {
      DateTime date = DateTime.parse(dateString);
      const months = [
        'Ocak',
        'Şubat',
        'Mart',
        'Nisan',
        'Mayıs',
        'Haziran',
        'Temmuz',
        'Ağustos',
        'Eylül',
        'Ekim',
        'Kasım',
        'Aralık'
      ];
      return months[date.month - 1];
    } catch (e) {
      return '';
    }
  }
}

File: src/services/CSVService.ts
import Papa from 'papaparse';
import {TufeDataModel} from '../models/TufeData';
import {GitHubCSVService} from './GitHubCSVService';

const MONTHS = [
  'Ocak',
  'Şubat',
  'Mart',
  'Nisan',
  'Mayıs',
  'Haziran',
  'Temmuz',
  'Ağustos',
  'Eylül',
  'Ekim',
  'Kasım',
  'Aralık',
];

export class CSVService {
  private static normalizeKey(name: string): string {
    return (name || '')
      .toLowerCase()
      .replace(/[\s,._-]+/g, '')
      .replace(/ı/g, 'i')
      .replace(/İ/g, 'i')
      .replace(/ö/g, 'o')
      .replace(/ü/g, 'u')
      .replace(/ğ/g, 'g')
      .replace(/ş/g, 's')
      .replace(/ç/g, 'c');
  }

  static async loadTufeData(): Promise<TufeDataModel[]> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'gruplaraylik.csv',
        false
      );

      const lines = csvData.split(/\r?\n/);

      const rows: any[][] = [];
      for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim();
        if (line) {
          try {
            const parsed = Papa.parse(line, {
              header: false,
              skipEmptyLines: true,
            });
            if (parsed.data.length > 0 && Array.isArray(parsed.data[0])) {
              rows.push(parsed.data[0] as any[]);
            }
          } catch (e) {
            // Parse hatası (sessiz)
          }
        }
      }

      if (rows.length === 0) {
        throw new Error('CSV dosyası boş');
      }

      const tufeDataList: TufeDataModel[] = [];

      // İlk satır header, ondan sonraki satırları işle
      for (let i = 1; i < rows.length; i++) {
        if (rows[i].length < 2) {
          continue;
        }

        const groupName = rows[i][1]?.toString().trim() || '';

        if (!groupName) {
          continue;
        }

        // Son sütundaki veriyi al (en güncel ay)
        const lastValue = rows[i][rows[i].length - 1];
        let changeRate = 0.0;

        if (typeof lastValue === 'number') {
          changeRate = lastValue;
        } else if (typeof lastValue === 'string') {
          try {
            changeRate = parseFloat(lastValue) || 0.0;
          } catch (e) {
            changeRate = 0.0;
          }
        }

        tufeDataList.push(new TufeDataModel(groupName, changeRate));
        console.log(`Eklendi: ${groupName} = ${changeRate}`);
      }

      console.log(`Toplam ${tufeDataList.length} veri eklendi`);

      // Verileri değişim oranına göre büyükten küçüğe sırala
      tufeDataList.sort((a, b) => b.changeRate - a.changeRate);

      return tufeDataList;
    } catch (e) {
      throw new Error(`CSV veri yükleme hatası: ${e}`);
    }
  }

  static getCurrentMonth(): string {
    const now = new Date();
    return MONTHS[now.getMonth()];
  }

  static async getMonthFromCSV(): Promise<string> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'gruplaraylik.csv',
        false
      );

      const lines = csvData.split(/\r?\n/);
      const rows: any[][] = [];

      for (const line of lines) {
        const trimmedLine = line.trim();
        if (trimmedLine) {
          try {
            const parsed = Papa.parse(trimmedLine, {
              header: false,
              skipEmptyLines: true,
            });
            if (parsed.data.length > 0 && Array.isArray(parsed.data[0])) {
              rows.push(parsed.data[0] as any[]);
            }
          } catch (e) {
            // Parse hatalarını sessizce atla
          }
        }
      }

      if (rows.length === 0 || rows[0].length < 3) {
        return this.getCurrentMonth();
      }

      // Header satırından son sütundaki tarihi al
      const lastDateString = rows[0][rows[0].length - 1]?.toString().trim() || '';

      // Tarih formatını parse et (2025-08-31 formatında)
      try {
        const lastDate = new Date(lastDateString);
        return MONTHS[lastDate.getMonth()];
      } catch (e) {
        return this.getCurrentMonth();
      }
    } catch (e) {
      return this.getCurrentMonth();
    }
  }

  static async getTufeMonthlyChange(): Promise<number> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'gruplaraylik.csv',
        false
      );

      const lines = csvData.split(/\r?\n/);
      const rows: any[][] = [];

      for (const line of lines) {
        const trimmedLine = line.trim();
        if (trimmedLine) {
          try {
            const parsed = Papa.parse(trimmedLine, {
              header: false,
              skipEmptyLines: true,
            });
            if (parsed.data.length > 0 && Array.isArray(parsed.data[0])) {
              rows.push(parsed.data[0] as any[]);
            }
          } catch (e) {
            // Parse hatalarını sessizce atla
          }
        }
      }

      if (rows.length === 0) {
        throw new Error('CSV dosyası boş');
      }

      // Find the "Web TÜFE" row
      for (let i = 1; i < rows.length; i++) {
        if (rows[i].length >= 2) {
          const groupName = rows[i][1]?.toString().trim() || '';
          if (groupName === 'Web TÜFE') {
            // Get the last column (latest monthly change)
            const lastValue = rows[i][rows[i].length - 1]?.toString().trim() || '';
            try {
              return parseFloat(lastValue) || 0.0;
            } catch (e) {
              return 0.0;
            }
          }
        }
      }

      return 0.0;
    } catch (e) {
      return 0.0;
    }
  }

  static async getAvailableDates(): Promise<string[]> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'gruplaraylik.csv',
        false
      );

      const lines = csvData.split(/\r?\n/);
      const rows: any[][] = [];

      for (const line of lines) {
        const trimmedLine = line.trim();
        if (trimmedLine) {
          try {
            const parsed = Papa.parse(trimmedLine, {
              header: false,
              skipEmptyLines: true,
            });
            if (parsed.data.length > 0 && Array.isArray(parsed.data[0])) {
              rows.push(parsed.data[0] as any[]);
            }
          } catch (e) {
            // Parse hatalarını sessizce atla
          }
        }
      }

      if (rows.length === 0 || rows[0].length < 3) {
        return [];
      }

      // Header satırından tarihleri al (2. sütundan itibaren)
      const dates: string[] = [];
      for (let i = 2; i < rows[0].length; i++) {
        const dateStr = rows[0][i]?.toString().trim() || '';
        if (dateStr) {
          dates.push(dateStr);
        }
      }

      return dates;
    } catch (e) {
      // Tarih listesi yükleme hatası (sessiz)
      return [];
    }
  }

  static async loadTufeDataForDate(
    selectedDate: string
  ): Promise<TufeDataModel[]> {
    try {
      const csvData = await GitHubCSVService.loadCSVFromGitHub(
        'gruplaraylik.csv',
        false
      );

      const lines = csvData.split(/\r?\n/);
      const rows: any[][] = [];

      for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim();
        if (line) {
          try {
            const parsed = Papa.parse(line, {
              header: false,
              skipEmptyLines: true,
            });
            if (parsed.data.length > 0 && Array.isArray(parsed.data[0])) {
              rows.push(parsed.data[0] as any[]);
            }
          } catch (e) {
            // Parse hatası (sessiz)
          }
        }
      }

      if (rows.length === 0) {
        throw new Error('CSV dosyası boş');
      }

      // Header satırından seçili tarihin sütun indeksini bul
      let dateColumnIndex = -1;
      for (let i = 2; i < rows[0].length; i++) {
        const dateStr = rows[0][i]?.toString().trim() || '';
        if (dateStr === selectedDate) {
          dateColumnIndex = i;
          break;
        }
      }

      if (dateColumnIndex === -1) {
        throw new Error(`Seçili tarih bulunamadı: ${selectedDate}`);
      }

      const tufeDataList: TufeDataModel[] = [];

      // İlk satır header, ondan sonraki satırları işle
      for (let i = 1; i < rows.length; i++) {
        if (rows[i].length < dateColumnIndex + 1) {
          continue;
        }

        const groupName = rows[i][1]?.toString().trim() || '';
        if (!groupName) {
          continue;
        }

        // Seçili tarih sütunundaki veriyi al
        const value = rows[i][dateColumnIndex];
        let changeRate = 0.0;

        if (typeof value === 'number') {
          changeRate = value;
        } else if (typeof value === 'string') {
          try {
            changeRate = parseFloat(value) || 0.0;
          } catch (e) {
            changeRate = 0.0;
          }
        }

        tufeDataList.push(new TufeDataModel(groupName, changeRate));
      }

      // Verileri değişim oranına göre büyükten küçüğe sırala
      tufeDataList.sort((a, b) => b.changeRate - a.changeRate);

      return tufeDataList;
    } catch (e) {
      throw new Error(`CSV veri yükleme hatası: ${e}`);
    }
  }

  static async loadContributionData(
    selectedDate?: string
  ): Promise<Array<{groupName: string; contribution: number; normalizedKey: string}>> {
    const csvData = await GitHubCSVService.loadCSVFromGitHub('katkıpayları.csv', true);
    const lines = csvData.split(/\r?\n/).filter((l) => l.trim().length > 0);
    if (lines.length < 2) {
      return [];
    }

    const rows: any[][] = [];
    for (const line of lines) {
      try {
        const parsed = Papa.parse(line, {header: false, skipEmptyLines: true});
        if (parsed.data.length > 0 && Array.isArray(parsed.data[0])) {
          rows.push(parsed.data[0] as any[]);
        }
      } catch {
        // ignore bad line
      }
    }
    if (rows.length < 2) return [];

    const header = rows[0];
    let dataRow: any[] | null = null;

    // If a date is provided, try to match that row
    if (selectedDate) {
      const match = rows.find((r, idx) => idx > 0 && `${r[0] || ''}`.trim() === selectedDate);
      if (match) dataRow = match;
    }

    // Fallback: last non-empty data row
    if (!dataRow) {
      for (let i = rows.length - 1; i >= 1; i--) {
        const r = rows[i];
        const hasValue = r.some((v, idx) => idx > 0 && v !== null && `${v}`.trim() !== '');
        if (hasValue) {
          dataRow = r;
          break;
        }
      }
    }

    if (!dataRow) return [];

    const items: Array<{groupName: string; contribution: number; normalizedKey: string}> = [];
    for (let col = 1; col < header.length; col++) {
      const groupName = header[col]?.toString().trim() || '';
      if (!groupName) continue;
      const rawVal = dataRow[col];
      let contribution = 0;
      if (typeof rawVal === 'number') contribution = rawVal;
      else if (typeof rawVal === 'string') {
        const parsed = parseFloat(rawVal.replace(',', '.'));
        if (!isNaN(parsed)) contribution = parsed;
      }
      items.push({
        groupName,
        contribution,
        normalizedKey: this.normalizeKey(groupName),
      });
    }
    return items;
  }

  static getMonthFromDate(dateString: string): string {
    try {
      const date = new Date(dateString);
      return MONTHS[date.getMonth()];
    } catch (e) {
      return '';
    }
  }
}

